Lib: Replace note links with relative paths in MD Exporter

vsimkus opened the pull request:
In my use case, I usually want to export my Joplin notes into a github repository and access them whenever necessary, even without my laptop or phone.

The current Markdown exporter in `InteropService_Exporter_Md.js` exports the notes as `.md` files, with resource/attachment links in the body replaced by relative paths to the `.md` file. However, links to linked notes are not handled. Hence, to complete the implementation it makes sense to implement the note link path replacement as well.

Same issue has been mentioned in:
* https://discourse.joplinapp.org/t/export-joplin-notes-to-markdown/1802
* https://discourse.joplinapp.org/t/markdown-export-links-invalid/2585

This PR replaces the linked note ids with relative paths to the `.md` file. The relative paths are URL-encoded such that white-space and other special characters are escaped.

Moreover, as before, the exporter handles duplicate note names and does not override existing note files by adding a counter to the filename. To ensure the correctness of the change, I've added a unit test suite for the `InteropService_Exporter_Md.js`.

laurent22 commented:
Thanks for the pull request and apologies for the late reply. There's a lot of code changes obviously and there's some of it I don't quite understand.

My expectation would be that it should be a two steps process:

- First: export all the notes as it's done now and keep a map of the note ID => note file path
- Second: Loop through the exported files and change the note IDs to the note file paths.

I don't quite understand how for example `findUniqueFilenameFromPathListAndFS` or `prepareForProcessingItemType` fits in there. There's also a change to the React Native shim which I guess should not be part of this PR since it's not possible to export from mobile.

vsimkus commented:
Hi thanks for the reply! 

The approach you describe is one way to do it and perhaps a little simpler. However, I see there could be a potentially significant performance overhead due to multiple read/writes to disk, if it has to loop through all the note files multiple times, especially on slower filesystems.

The process in the current implementation:
1. Create FS paths for all notes to be exported.
2. Loop through the notes to be exported, replace the note IDs in each note with note FS paths from step 1 (the same way resource IDs are replaced). Write note to disk at the FS path from step 1.

I created the `prepareForProcessingItemType` as a generic interface for any item type, if any preprocessing is necessary for a particular exporter. In MD exporter, its purpose is to create the FS paths for each note in the export list. The note paths are then stored in the context as `notePaths` for use in step 2.

Inside `prepareForProcessingItemType`, `findUniqueFilenameFromPathListAndFS` is called to create unique filenames for each note. When creating a filename it checks that the filename does not already exist in the FS and also that it has not been allocated to another note in the current list of exported items (since Joplin allows notes with the same name). The counter increment rule is the same as in `fs-driver-base.js/findUniqueFilename`.

Changes to RN shim: I've added the `readDir` function to maintain the same API for both node and RN FS interfaces. It's best practice to maintain the same functionality in both abstractions. I am happy to remove the function from `fs-driver-rn.js` if you think it shouldn't be there. 

Happy Christmas and the coming New Year.

laurent22 commented on ReactNativeClient/lib/fs-driver-node.js:
```diff
@@ -160,6 +160,15 @@ class FsDriverNode extends FsDriverBase {
+			throw this.fsErrorToJsError_(error, path);
+		}
+	}
+
```
To get the files inside a directory, please use `readDirStats()`.

vsimkus replied:
Ah yes, I didn't notice this function. Maybe the naming of it could just be `readDir()`, or `readDirFiles()`. It's not immediately clear what "stats" mean in the context of a directory.

laurent22 commented on ReactNativeClient/lib/fs-driver-rn.js:
```diff
@@ -107,6 +107,10 @@ class FsDriverRN extends FsDriverBase {
+	async readDir(path) {
+		return RNFS.readDir(path).map(e => e.name);
+	}
+
```
As `readDirStats` can be used, we won't need this function anymore.

laurent22 commented on ReactNativeClient/lib/services/InteropService_Exporter_Md.js:
```diff
@@ -31,36 +31,105 @@ class InteropService_Exporter_Md extends InteropService_Exporter_Base {
+			if (counter >= 1000) nameToTry = `${nameNoExt} (${new Date().getTime()})${extension}`;
+			if (counter >= 10000) throw new Error('Cannot find unique title');
+		}
+	}
```
Is it not possible to use the existing logic with `friendlySafeFilename()` / `findUniqueFilename()` here? We could add a second "existingNames" parameter to `findUniqueFilename(name, existingNames = null)` and if presents it is used to check for existing filenames too. I'm not entirely sure it will replicate all the required logic so please confirm.

vsimkus replied:
Modified the existing `findUniqueFilename()` by adding a "reservedNames" parameter (I think it's slightly better name than "existingNames" because the files in the `reservedNames` are not yet created but are only reserved).

laurent22 commented:
Thanks for the explanation @vsimkus. I've just added a few more comments but otherwise it's nearly ready to merge. Thanks also for the extensive test units, they are very well done and are going to be useful to keep this complex logic stable.

vsimkus commented:
I'm glad the unit tests are appreciated :) 

I've made your suggested changes - the change to `findUniqueFilename()` was very trivial. Also removed the `readDir()` methods I've added beforehand.

laurent22 commented:
That's all good now, many thanks for the pull request @vsimkus!
